lib/logstash/outputs/tcp.rb

```diff
@@ -204,7 +176,38 @@ def server?
   end # def server?
 
   public
-  def receive(event)
-    @codec.encode(event)
-  end # def receive
+  def multi_receive_encoded(encoded)
+    if server?
+      @server_mutex.synchronize do
+        @client_threads.each do |client_thread|
+           encoded.each do |event,data|
+             client_thread[:client].write(data)
+           end
+        end
+        @client_threads.reject! {|t| !t.alive? }
+      end
+    else
+      client_socket = nil
+      begin
+        client_socket = connect unless client_socket
```

Comment:
Not sure if you already planned this or not, but just copying the code from the original here is problematic. This will always reconnect on each batch!

We should have a client socket per worker thread. Code like this might be good.

`client_socket = Thread.current[:client_socket] ||= connect`